<a href="https://colab.research.google.com/github/calvinarieri/machine-learning-course/blob/main/AI_Traffic_Light_Control_Project.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# AI/ML-Based Intelligent Traffic-Light Control

## Project Problem and Intended AI/ML Outcome

Urban traffic congestion causes increased vehicle waiting times, longer queues,
higher fuel consumption, increased travel time, and environmental pollution.
Many traffic-light systems use fixed timing schedules and may not respond
effectively to changing traffic conditions.

This project investigates how Artificial Intelligence (AI) and Machine Learning
(ML) can be used to predict traffic conditions and support intelligent
traffic-light control.

The intended AI/ML outcome is to develop a model that can use historical and
real-time traffic information to predict traffic flow or congestion. These
predictions could then be used by an intelligent traffic-control system to
dynamically adjust traffic-light timing in order to reduce waiting time,
queue lengths, unnecessary stops, and overall travel time.

For this assignment, the machine-learning component will focus on predicting
traffic volume/flow from available traffic and time-related features.


## Data Requirements

The proposed intelligent traffic-light system would ideally require both
historical and real-time traffic data.

Useful data would include:

- Vehicle counts or traffic volume
- Vehicle speed
- Traffic density
- Road occupancy
- Queue length
- Date and time
- Day of the week
- Traffic-light phase and timing
- Vehicle turning movements
- Weather conditions
- Accidents and road closures
- Pedestrian and public-transport demand

For this project, the available dataset will be used to develop a model for
predicting traffic flow/volume using traffic and time-related variables.


In [1]:
import pandas as pd

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/00492/Metro_Interstate_Traffic_Volume.csv.gz"

df = pd.read_csv(url)

print("Dataset loaded successfully!")
print("Number of rows:", df.shape[0])
print("Number of columns:", df.shape[1])


Dataset loaded successfully!
Number of rows: 48204
Number of columns: 9


In [4]:
df.head()

,holiday,temp,rain_1h,snow_1h,clouds_all,weather_main,weather_description,date_time,traffic_volume
0,NaN,288.28,0.0,0.0,40,Clouds,scattered clouds,2012-10-02 09:00:00,5545
1,NaN,289.36,0.0,0.0,75,Clouds,broken clouds,2012-10-02 10:00:00,4516
2,NaN,289.58,0.0,0.0,90,Clouds,overcast clouds,2012-10-02 11:00:00,4767
3,NaN,290.13,0.0,0.0,90,Clouds,overcast clouds,2012-10-02 12:00:00,5026
4,NaN,291.14,0.0,0.0,75,Clouds,broken clouds,2012-10-02 13:00:00,4918


In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 48204 entries, 0 to 48203
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   holiday              61 non-null     object 
 1   temp                 48204 non-null  float64
 2   rain_1h              48204 non-null  float64
 3   snow_1h              48204 non-null  float64
 4   clouds_all           48204 non-null  int64  
 5   weather_main         48204 non-null  object 
 6   weather_description  48204 non-null  object 
 7   date_time            48204 non-null  object 
 8   traffic_volume       48204 non-null  int64  
dtypes: float64(3), int64(2), object(4)
memory usage: 3.3+ MB


In [6]:
df.describe()

,temp,rain_1h,snow_1h,clouds_all,traffic_volume
count,48204.000000,48204.000000,48204.000000,48204.000000,48204.000000
mean,281.205870,0.334264,0.000222,49.362231,3259.818355
std,13.338232,44.789133,0.008168,39.015750,1986.860670
min,0.000000,0.000000,0.000000,0.000000,0.000000
25%,272.160000,0.000000,0.000000,1.000000,1193.000000
50%,282.450000,0.000000,0.000000,64.000000,3380.000000
75%,291.806000,0.000000,0.000000,90.000000,4933.000000
max,310.070000,9831.300000,0.510000,100.000000,7280.000000


## Dataset

The dataset selected for this project is the **Metro Interstate Traffic Volume**
dataset from the UCI Machine Learning Repository.

The dataset contains hourly traffic observations from Interstate 94 in the
Minneapolis-St. Paul area from 2012 to 2018. It contains traffic volume,
weather conditions, holidays, and date/time information.

The dataset contains 48,204 observations and 9 variables. The main target
variable is `traffic_volume`, which represents the hourly number of vehicles
recorded at the monitoring station.

### Dataset Source

UCI Machine Learning Repository:
Metro Interstate Traffic Volume

The dataset is suitable for this project because traffic volume can be used as
an indicator of traffic demand. Predicting traffic volume can help an
intelligent traffic-control system anticipate changing traffic conditions and
make better traffic-light timing decisions.


## Potential Inputs (Features) and Output (Target)

### Input Features

The potential input features are:

- Holiday information
- Temperature
- Rainfall
- Snowfall
- Cloud coverage
- Weather conditions
- Weather description
- Hour of the day
- Day of the week
- Month

The `date_time` variable will be processed to extract useful time-based
features such as hour, day of the week, and month.

### Output/Target

The target variable is:

**`traffic_volume`**

This represents the number of vehicles recorded during an hour.

The machine-learning model will learn the relationship between the input
features and traffic volume so that it can predict future traffic demand.

These predictions could eventually be used by an intelligent traffic-light
controller to adjust signal timing according to expected traffic conditions.


## Proposed ML Workflow

The proposed machine-learning workflow is:

```text
┌──────────────────────────────┐
│ Historical Traffic Dataset   │
│ Traffic + Weather + Time Data│
└──────────────┬───────────────┘
               │
               ▼
┌──────────────────────────────┐
│ Data Preprocessing           │
│ - Handle missing values      │
│ - Process date/time           │
│ - Encode categorical data    │
└──────────────┬───────────────┘
               │
               ▼
┌──────────────────────────────┐
│ Feature Selection/Engineering│
│ - Hour                       │
│ - Day of week                │
│ - Month                      │
│ - Weather variables          │
│ - Holiday information        │
└──────────────┬───────────────┘
               │
               ▼
┌──────────────────────────────┐
│ Train ML Model               │
│                              │
│ Input: Traffic-related data  │
│ Target: Traffic volume       │
└──────────────┬───────────────┘
               │
               ▼
┌──────────────────────────────┐
│ Model Evaluation             │
│ - MAE                        │
│ - RMSE                       │
│ - R²                         │
└──────────────┬───────────────┘
               │
               ▼
┌──────────────────────────────┐
│ Predict Future Traffic       │
│ Volume/Traffic Conditions    │
└──────────────┬───────────────┘
               │
               ▼
┌──────────────────────────────┐
│ Intelligent Traffic Control  │
│                              │
│ Use predictions to support   │
│ dynamic traffic-light timing │
└──────────────────────────────┘


## Project Summary

This project investigates the use of Artificial Intelligence and Machine
Learning to improve urban traffic management.

The selected dataset contains traffic volume and related weather, holiday,
and time information. The proposed ML model will use these variables to
predict traffic volume.

Accurate traffic-volume predictions could support an intelligent traffic-light
control system by allowing signal timings to respond to expected traffic
conditions rather than relying only on fixed schedules.

The ultimate goal is to reduce traffic congestion, vehicle waiting time,
queue lengths, unnecessary stops, and overall travel time.